geomesa-features/geomesa-feature-avro-complex/src/main/scala/org/locationtech/geomesa/features/avro/complex/AvroCodecs.scala

```diff
@@ -0,0 +1,697 @@
+import org.opengis.feature.`type`.{PropertyDescriptor, _}
+import org.opengis.feature.{Attribute, ComplexAttribute, Property}
+import org.opengis.filter.identity.FeatureId
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+object AvroCodecs {
+
+  val primitiveTypes =
+    List(
+      classOf[String],
+      classOf[java.lang.Integer],
+      classOf[Int],
+      classOf[java.lang.Long],
+      classOf[Long],
+      classOf[java.lang.Double],
+      classOf[Double],
+      classOf[java.lang.Float],
+      classOf[Float],
+      classOf[java.lang.Boolean],
+      classOf[Boolean]
+    )
+
+  abstract case class AvroCodec[T](schema:Schema) {
+    val clazz: Class[T]
+    def enc(t: T, enc: Encoder): Unit
+    def dec(dec: Decoder): T
+  }
+
+  trait CodecRegistry {
+    protected val registry = mutable.Map.empty[Class[_],mutable.LinkedHashMap[Schema.Type,AvroCodec[_]]]
+    protected val cache = mutable.Map.empty[Class[_], Option[mutable.LinkedHashMap[Schema.Type,AvroCodec[_]]]]
+
+    def bind[T](encMethod: (Encoder,T)=>Unit, decMethod: (Decoder)=>T, schema: Schema)(implicit tag: ClassTag[T]) = {
+      val codec = new AvroCodec[T](schema) {
+        val clazz = tag.runtimeClass.asInstanceOf[Class[T]]
+        override def enc(t: T, out: Encoder): Unit = encMethod(out, t)
+        override def dec(in: Decoder): T = decMethod(in)
+      }
+      registry.getOrElseUpdate(tag.runtimeClass,mutable.LinkedHashMap.empty)(schema.getType) = codec
+      codec
+    }
+
+    val JBOOLEAN = classOf[java.lang.Boolean]
+    val JBYTE = classOf[java.lang.Byte]
+    val JCHAR = classOf[java.lang.Character]
+    val JSHORT = classOf[java.lang.Short]
+    val JINT = classOf[java.lang.Integer]
+    val JLONG = classOf[java.lang.Long]
+    val JFLOAT = classOf[java.lang.Float]
+    val JDOUBLE = classOf[java.lang.Double]
+
+    def normalizePrimitive(clazz:Class[_]) = clazz match {
+      case JBOOLEAN | java.lang.Boolean.TYPE => classOf[Boolean]
+      case JBYTE | java.lang.Byte.TYPE => classOf[Byte]
+      case JCHAR | java.lang.Character.TYPE => classOf[Char]
+      case JSHORT | java.lang.Short.TYPE => classOf[Short]
+      case JINT | java.lang.Integer.TYPE => classOf[Int]
+      case JLONG | java.lang.Long.TYPE => classOf[Long]
+      case JFLOAT | java.lang.Float.TYPE => classOf[Float]
+      case JDOUBLE | java.lang.Double.TYPE => classOf[Double]
+      case x => x
+    }
+
+    def findAll[T](clazz:Class[T]): Option[mutable.LinkedHashMap[Schema.Type, AvroCodec[_]]] =
+      cache.getOrElseUpdate(clazz, registry.get(normalizePrimitive(clazz))
+        .orElse(Option(clazz.getSuperclass).flatMap(findAll(_)))
+        .orElse {
+          for (ifc <- clazz.getInterfaces; map <- findAll(ifc))
+            return Some(map)
+          None
+        }
+      )
+
+    def find[T](clazz:Class[T], s:Schema.Type): Option[AvroCodec[_>:T]] =
+      findAll(clazz).flatMap(m=>m.get(s)).asInstanceOf[Option[AvroCodec[_>:T]]]

```

Comment:
Ah, ok, that might be why... you are trying to expose AvroCodec if you find them...